include/xtensor-sparse/xmap.hpp

```diff
@@ -0,0 +1,285 @@
+        return *this;
+    }
+    
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::reshape_impl(S&& shape, std::false_type /* is unsigned */)
+    {
+        if (compute_size(shape) != this->size())
+        {
+            XTENSOR_THROW(std::runtime_error, "Cannot reshape with incorrect number of elements. Do you mean to resize?");
+        }
+
+        update_keys(std::forward<S>(shape));
+        m_shape = xtl::forward_sequence<shape_type, S>(shape);
+    }
+
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::reshape_impl(S&& _shape, std::true_type /* is signed */)
+    {
+        using value_type = typename std::decay_t<S>::value_type;
+        auto new_size = compute_size(_shape);
+        if (this->size() % new_size)
+        {
+            XTENSOR_THROW(std::runtime_error, "Negative axis size cannot be inferred. Shape mismatch.");
+        }
+        std::decay_t<S> shape = _shape;
+        value_type accumulator = 1;
+        std::size_t neg_idx = 0;
+        std::size_t i = 0;
+        for(auto it = shape.begin(); it != shape.end(); ++it, i++)
+        {
+            auto&& dim = *it;
+            if(dim < 0)
+            {
+                XTENSOR_ASSERT(dim == -1 && !neg_idx);
+                neg_idx = i;
+            }
+            accumulator *= dim;
+        }
+        if(accumulator < 0)
+        {
+            shape[neg_idx] = static_cast<value_type>(this->size()) / std::abs(accumulator);
+        }
+        else if(this->size() != new_size)
+        {
+            XTENSOR_THROW(std::runtime_error, "Cannot reshape with incorrect number of elements. Do you mean to resize?");
+        }
+        
+        update_keys(shape);
+        m_shape = xtl::forward_sequence<shape_type, S>(shape);
+    }
+
+    template <class T>
+    template <class S>
+    inline void xmap_container<T>::update_keys(S&& shape)
+    {
+        container_type new_data;
+        for(auto& c: m_data)
+        {
+            auto& old_key = c.first;
+
+            // compute the offset for the old_key
+            size_type index = 0;
+            size_type accumulator = 1;
+            for (std::size_t i = m_shape.size(); i != 0; --i)
+            {
+                index += old_key[i - 1]*accumulator;
+                accumulator *= m_shape[i - 1];
+            }
+
+            // compute the new key from the offset and the new shape
+            index_type new_key = xtl::make_sequence<index_type>(shape.size());
+            accumulator = 1;
+            for(std::size_t i = 0; i < shape.size(); ++i)
+            {
+                accumulator *= static_cast<size_type>(shape[i]);
+            }
+            
+            for(std::size_t i = 0; i < shape.size(); ++i)
+            {
+                accumulator /= static_cast<size_type>(shape[i]);
+                new_key[i] = index/accumulator;
+                index -= new_key[i]*accumulator;
```

Comment:
What about storing the corresponding strides, and reuse element_offset and unravel_index?

Reply:
It's a good idea. I didn't know these functions!